File: src/main/java/graphics/LevelSelectPane.java
package graphics;

import graphics.controls.LevelDiagram;
import io.Level;
import io.Reader;
import javafx.fxml.*;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.*;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;

import java.net.URL;
import java.util.List;
import java.util.ResourceBundle;

/**
 * Created by Mitch on 5/21/2017.
 */
public class LevelSelectPane implements Initializable {

    @FXML private TreeView<Node> tree_view;
    @FXML private VBox level_details;
    @FXML private Label level_name;
    @FXML private Label level_desc;
    @FXML private StackPane level_diagram;
    @FXML private Label level_completion_info;

    private LevelDiagram diagram = new LevelDiagram(300);

    @FXML public void nav_title(){
        Main.set_screen("TitlePane");
    }

    public void set_level(Level level){
        diagram.set_level(level);
        level_details.setVisible(true);
        level_name.setText(level.name());
        level_desc.setText(level.instruction_text());
        level_completion_info.setText(
                "Completed with size 8\nMin size "+level.min_size()
        );

        System.out.println(level_desc.getWidth());
    }

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        TreeItem<Node> root = new TreeItem<>(new Label("Levels"));

        for (int i = 0; i < Reader.LEVELS_JAVA().size(); i++){
            TreeItem<Node> level_set_root = new TreeItem<>(new Label(i+1 +". "+ Reader.LEVEL_SET_NAMES_JAVA().get(i)));
            List<Level> level_set = Reader.LEVELS_JAVA().get(i);
            for(int j = 0; j < Reader.LEVELS_JAVA().get(i).size(); j++){
                Label l = new Label(String.format("%d.%d. %s", i+1, j+1, level_set.get(j).name()));
                final int I = i, J = j;
                l.setOnMouseClicked(e ->
                    this.set_level(Reader.LEVELS_JAVA().get(I).get(J))
                );
                level_set_root.getChildren().add(new TreeItem<>(l));
            }
            root.getChildren().add(level_set_root);
        }

        tree_view.setRoot(root);
        level_diagram.getChildren().add(diagram);
    }
}

File: src/main/scala/graphical/screens/LevelSelectPane.scala
package main.scala.graphical.screens

import javafx.event.Event
import javafx.scene.input.MouseEvent

import core.types.Coord
import graphical.screens.TitlePane
import io.{Level, Reader}
import main.scala.graphical.Main
import main.scala.graphical.controls.LevelDiagram

import scalafx.scene.control._
import scalafx.scene.layout.{HBox, Priority, VBox}

/**
  * Created by Mitch on 7/30/2017.
  */
object LevelSelectPane extends HBox {
  private val tree_view             = init_tree()
  private val level_name            = new Label("NAME OF LEVEL") {id = "level_name"}
  private val level_desc            = new Label("text goes here") {id = "level_desc"}
  private val level_diagram         = new LevelDiagram(300)
  private val level_completion_info = new Label("completion info") {id = "level_completion_info"}
  private val level_details         = new VBox {
    children = Seq(
      level_name,
      level_desc,
      level_diagram,
      level_completion_info,
      new Button(text = "Attempt")
    )
    id = "level_details"
    styleClass = Seq("bordered")
    hgrow = Priority.Always
    visible = false
  }

  children = Seq (
    new VBox {
      children = Seq(
        tree_view,
        new Button {
          text = "Menu"
          maxWidth = Double.MaxValue
          onMouseClicked = _ => Main.set_screen(TitlePane)
        }
      )
      styleClass = Seq("bordered")
    },
    level_details
  )
  id = "root"
  stylesheets = Seq(
    "@../../css/all-panes.css",
    "@../../css/level-select-pane.css"
  )

  init_tree()

  def set_level(level: Level): Unit = {
    level_diagram.set_level(level)
    level_details.setVisible(true)
    level_name.setText(level.name)
    level_desc.setText(level.instruction_text)
    level_completion_info.setText("Completed with size 8\nOptimal size " + level.optimal_size)
  }
  private def init_tree(): TreeView[LevelCell] =
    new TreeView[LevelCell] {
      root = new TreeItem[LevelCell] {
        children = for (i <- Reader.LEVEL_SETS.indices) yield
        new TreeItem[LevelCell] {
          children = for (j <- Reader.LEVEL_SETS(i).levels.indices)
            yield new TreeItem(new LevelCell(i, j))
        }
      }
      id = "tree_view"
      vgrow = Priority.Always
      showRoot = false
    }

  private class LevelCell(i: Int, j: Int) extends TreeCell[LevelCell] {

    //s"${i+1}. " + Reader.LEVEL_SET_NAMES(i)
    onMouseClicked = _ => println(s"($i, $j)")

    addEventHandler[MouseEvent](MouseEvent.MOUSE_CLICKED, _ => {})

    override def toString(): String = "%d.%d. %s".format(i+1, j+1, Reader.level(i, j).get.name)
  }
}